### ISSUE TITLE
[feature] Allow setting of vendor specific information for generated CMake presets

### ISSUE THREAD

**Issue by WernerWessely** [conan-io/conan#12963]
### What is your suggestion?

Hello, and thanks for your great work on Conan!

We really like the automatic generation of CMake presets feature and want to make it a fixed part of our (future) workflow.

Here is our use-case: Many in our teams use Clion for their C/C++ development. Clion has ever improving support for presets and they use the 'vendor' section of the configure presets for the user to specify which 'Clion toolchain' he wants to use, this might look something like:
```
    {
      "name": "my_preset",
      "vendor": {
        "jetbrains.com/clion": {
          "toolchain": "my-docker-toolchain"
        }
      }
    }
```
We currently use this feature to configure/build/test/debug builds in Docker containers or WSL. The only issue is, that we cannot set this information in any way when generating the presets with Conan. We would have to patch the generated presets or create some new ones which are derived from the generated ones.

Perhaps Conan could provide a way to set this section for the generated configure/build/test presets? 

### Have you read the CONTRIBUTING guide?

- [x] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @WernerWessely 

We are considering a different approach for this issue, in https://github.com/conan-io/conan/pull/12932. Basically, instead of adding more information to the Conan generated presets, users can have their own presets and ``inherits`` the Conan generated ones. That would allow for more flexibility. If you can please check the PR, and specially the unittest that covers the case (it should be pretty explicit, with the ``"include ConanPresets.json"``. Thanks!

**Comment by WernerWessely** [comment]
@memsharded 
I will take a look today.

**Comment by WernerWessely** [comment]
@memsharded So I tried to understand the code and tests (never looked at Conan code before).

The only major downside I see with this general approach is this: I will never be able to use any of the presets in my project until I have installed and generated ALL of the presets that all of my self written presets depend on. These can come from many different profiles.

Whether if I provide a way to inject any information into the generated presets, I have max. flexibility because the project will only see the preset information generated by conan and nothing else.

Perhaps something like this:

`tools.cmake.cmaketoolchain:preset_includes` list of preset files to include in generators/CMakePresets.json
`tools.cmake.cmaketoolchain:configure_preset_inherits`: name of preset that generated configure preset should inherit from.

This is just an example approach, there is probably something smarter.

The option to inject into the generated presets so that you only have to use the information generated by Conan would insure that you can always use your project's presets even if you only installed the profile/profiles that you need in order to work. Imagine you have some profiles which you can't even install on your windows machine because they only work on linux. Then your self written presets cannot be used if they extend the ones from linux as well.

I know it's kinda complicated but maybe it's worth thinking about some more?

**Comment by WernerWessely** [comment]
Actually, `tools.cmake.cmaketoolchain:configure_preset_inherits` should probably be a list as well. It's really a pity that presets cannot inherit from presets defined in files which they themselves are being included in.

**Comment by WernerWessely** [comment]
Mmm actually I just thought of another way:

`tools.cmake.cmaketoolchain.presets_includes=[cmake/presets/my-presets.json]`

cmake/presets/my-presets.json:
```
{
  "includes": ["<some-path>/generators/CMakePresets.json"],
  "configurePresets": [
    {
      "name": "my-preset",
      "inherits": ["conan-generated-preset"]
    }
  ]
}
```
CMakeUserPresets.json generated by conan:
```
{
  "includes": [
      "cmake/presets/my-presets.json",
      "<some-path>/generators/CMakePresets.json"
  ]
}
```

Kinda finicky because I have to hardcode the path to the generated file in my-presets.json but ok...

**Comment by memsharded** [comment]
> The only major downside I see with this general approach is this: I will never be able to use any of the presets in my project until I have installed and generated ALL of the presets that all of my self written presets depend on. These can come from many different profiles.

Not really. Conan will generate empty presets to inherit for all the configurations that you are not generating with Conan. I already hit this, and I implemented this, so the test in the PR can actually build both in Linux and Windows without needing to install the other configurations.

**Comment by memsharded** [comment]
> Whether if I provide a way to inject any information into the generated presets, I have max. flexibility because the project will only see the preset information generated by conan and nothing else.

There is also another perspective: There will be users that they have their own presets, and they will want to keep using their presets. No way they will leverage the creation of those presets or the inclusion of those presets by Conan. Plus, if they have already the CMakePresets.json in the root of the project, and possibly a CMakeUserPresets.json, then Conan will not be able to overwrite it.  This is what many users call being "intrusive", requiring them to change their flows and files.

So we are looking for something that is good for the different use cases, and the flow of allowing users to use their own CMakePresets file in their repo is necessary.

**Comment by WernerWessely** [comment]
> Conan will generate empty presets to inherit for all the configurations that you are not generating with Conan.

What do you mean? How does conan know which configurations I need for all my different hand written presets?

**Comment by WernerWessely** [comment]
> There is also another perspective: There will be users that they have their own presets, and they will want to keep using their presets. No way they will leverage the creation of those presets or the inclusion of those presets by Conan. Plus, if they have already the CMakePresets.json in the root of the project, and possibly a CMakeUserPresets.json, then Conan will not be able to overwrite it.

Yes, I have nothing against the `user_presets_path` feature, that seems perfectly useful.

**Comment by memsharded** [comment]
> What do you mean? How does conan know which configurations I need for all my different hand written presets?

It will use the ``CMakeToolchain.preset_prefix`` and check all your ``inherits`` in your existing presets, all those starting with the defined prefix, will be created as empty if not actually installed by Conan.

**Comment by WernerWessely** [comment]
Right, I skipped that part in the code sry 🥲. Yeah that actually seems like a pretty good solution.

**Comment by WernerWessely** [comment]
That would mean that I would have to install at least ONE profile before I can use any of my handwritten presets though right? Because I have to include the `ConanPresents.json` (or whatever) file in either CMakePresets.json or CMakeUserPresets.json, wherever I have written them.

**Comment by memsharded** [comment]
> That would mean that I would have to install at least ONE profile before I can use any of my handwritten presets though right? Because I have to include the ConanPresents.json (or whatever) file in either CMakePresets.json or CMakeUserPresets.json, wherever I have written them.

yes, it is necessary to install at least some configuration.

**Comment by WernerWessely** [comment]
Mmm that seems... like it will be annoying in the future. Image you want to have a preset to build the docs, something that has no relevance to conan. You cannot use a preset unless you create a doc profile which may not make any sense at all... I guess thats the way it is then...

**Comment by memsharded** [comment]
> Mmm that seems... like it will be annoying in the future. Image you want to have a preset to build the docs, something that has no relevance to conan. You cannot use a preset unless you create a doc profile which may not make any sense at all... I guess thats the way it is then...

No need to create a doc profile, but just use any other profile, like the default one or something. But yes, I see no other way, if users have their own Presets files they want to use, the only way to extend them with Conan provided information is to include + inherits, and if CMake decides that all inherits must exist at the time of parsing, not at the time of usage, there is little else we can do? Maybe report it to CMake and request that ``inherits`` are lazy and only active if used?

**Comment by WernerWessely** [comment]
Yeah maybe down the road if it becomes a real issue for people... Anyway yeah I guess we will work with this feature then. Thanks a lot for making it happen!
